nonthermal.cc

```diff
@@ -467,7 +467,64 @@ void init(const int my_rank, const int ndo_nonempty) {
     fflush(nonthermalfile);
   }
 
+  int LIST_LENGTH = MAX_NT_EXCITATIONS_STORED;
+#ifdef MPI_ON
+
+  MPI_Win nt_solution_window;
+
+  // We want to keep the nt_solution_struct in local but the nt_excitation_struct in shared memory
+  int rank, shm_rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+  // Get the rank of the shared memory communicator
+  MPI_Comm shmcomm;
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &shmcomm);
+
+  // Rank within the shared memory communicator
+  MPI_Comm_rank(shmcomm, &shm_rank);
+
+  // The max size of the vector is set to MAX_NT_EXCITATIONS_STORED hence the memory allocation is done for that size
+  int list_size = grid::get_npts_model() * sizeof(nt_excitation_struct) * LIST_LENGTH;
+
+  // Now add a printout telling us how much memory we are using / going to use
+  printout("[info] mem_usage: storing non-thermal frac_excitations_list for all allocated cells occupies %.3f MB\n",
+           list_size / 1024. / 1024.);
+
+  // MPI_Aint is the type of a variable able to contain a memory address.
+  MPI_Aint shared_mem_size = (shm_rank == 0) ? list_size : 0;
+
+  // Allocate shared memory for frac_excitations_list on local rank 0
```

Comment:
This might be ok (and simplifies the code), but I always worried that some clusters would failed to account for the memory properly when everything is assigned to node_rank 0. In other parts of the code, we distribute the array as evenly as possible.

Reply:
So the method of allocation onto rank 0 can cause problems because it might unevenly allocate memory on the node ? Subsequently what is the method for distributing the array evenly (If you can point to a similar example I am happy to try and implement that method)

Reply:
I never confirmed it, but we often we request a certain amount of memory "per task" and I'm not sure if one of the tasks exceeding that limit will cause a problem (or whether the memory of all tasks on a node is correctly averaged out).

If you want more examples, just search the code for MPI_Win_allocate_shared.